### PR TITLE
Some optimizations

### PR DESCRIPTION
**Pull request by beru**
Hi, this PR adds some optimizations.
About 5% faster than master branch with `example_cifar_train` program.

### CODE REVIEW

File: tiny_dnn/util/product.h [hunk 1]
```diff
@@ -217,218 +257,357 @@ struct double_avx {
...
-                           typename T::value_type *dst) {
-  for (std::size_t i = 0; i < size / T::unroll_size; i++) {
-    typename T::register_type d = T::load(&dst[i * T::unroll_size]);
-    typename T::register_type s = T::load(&src[i * T::unroll_size]);
-    T::store(&dst[i * T::unroll_size], T::add(d, s));
+void fill(T *dst, size_t size, T value) {
+// On recent x86/x64 processors, stos instructions are faster
+// than SSE/AVX store instructions and recent compilers generates
+// stos instructions with std::fill
+#if 1
```
Comment (Randl) [inline tiny_dnn/util/product.h]: What's the point of this?

**Comment by Randl** [comment]
Please, rebase on master

File: tiny_dnn/util/product.h (hunk 1 above)
Comment (beru) [inline tiny_dnn/util/product.h]: This comment explains why simple `std::fill` is used over inhouse `vectorize::detail::fill_impl`.
Simple unrolled SSE/AVX store instructions isn't the fastest with recent x86/x64 processors.

With Visual Studio, there're intrinsic functions for `stosd` instruction but I can't find equivalent once for gcc/clang. Fortunately, compilers follow `64-ia-32-architectures-optimization-manual.pdf` so they produce right instructions.

**Comment by beru** [comment]
Rebased.

File: tiny_dnn/util/product.h (hunk 1 above)
Comment (Randl) [inline tiny_dnn/util/product.h]: I meant, why do we need `#if 1`?

Comment (beru) [inline tiny_dnn/util/product.h]: A huge advantage of using `#if 1` is I can switch the code by change a letter.
But a lot of people blamed me for leaving dead code and some reasons.

Comment (Randl) [inline tiny_dnn/util/product.h]: But is it likely that you need to switch it on fly often? If not, isn't it better to just remove it and recover from history if needed?

Comment (beru) [inline tiny_dnn/util/product.h]: No, I don't switch it often. If we remove the code, I think we will be very likely to forget the reason why it is buried.

Comment (Randl) [inline tiny_dnn/util/product.h]: So maybe leave to comment but remove the code?

Comment (beru) [inline tiny_dnn/util/product.h]: I removed the code with recent commit. We can recover it later if we find it useful.

**Comment by edgarriba** [comment]
LGTM

@beru hope you know what you do, I don't speak avx :smile:
Again, we should include some regression tests to see this kind of improvements

**Pull request merged**